File: three-sum/python/three-sum.py
def threeSum(self, nums):
    """
    :type nums: List[int]
    :rtype: List[List[int]]
    """
    result = []
    nums.sort()
    for i in range(len(nums)):
        if nums[i] > 1:
            break
        if i == 0 or nums[i - 1] != nums[i]:
            self.twoSum(i, nums, result)
            
    return result
    
def twoSum(self, i, nums, result):
    low, high = i + 1, len(nums) - 1
        
    while low < high:
        if nums[i] + nums[low] + nums[high] < 0:
            low += 1
        elif nums[i] + nums[low] + nums[high] > 0:
            high -= 1
        else:
            result.append([nums[i], nums[low], nums[high]])
            low += 1
            high -= 1
                
            while low < high and nums[low - 1] == nums[low]:
                low += 1

File: three-sum/java/three-sum.java
public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList();
        Arrays.sort(nums);
        for(int i = 0; i < nums.length; i++){
                if(nums[i] > 1) break;
                        if(i == 0 || nums[i - 1] != nums[i]){
                        twoSum(i, nums, result);
                        }
                }
        return result;
}
    
private void twoSum(Integer i, int[] nums, List<List<Integer>> result){
        Integer low = i + 1, high = nums.length - 1;
        while(low < high){
                if(nums[i] + nums[low] + nums[high] < 0){
                        low++;
                } else if(nums[i] + nums[low] + nums[high] > 0){
                        high--;
                } else {
                        result.add(Arrays.asList(nums[i], nums[low++], nums[high--]));
                        while(low < high && nums[low - 1] == nums[low]) low++;
                }
        }
}